mhs/outbound/outbound/request/synchronous/handler.py

```diff
@@ -132,30 +134,47 @@ def _should_invoke_sync_async_workflow(self, interaction_config, sync_async_head
+    async def _invoke_sync_async(self, from_asid, message_id, correlation_id, interaction_details, body, async_workflow):
         sync_async_workflow: workflow.SyncAsyncWorkflow = self.workflows[workflow.SYNC_ASYNC]
-        status, response, wdo = await sync_async_workflow.handle_sync_async_outbound_message(message_id,
+        status, response, wdo = await sync_async_workflow.handle_sync_async_outbound_message(from_asid,
+                                                                                             message_id,
                                                                                              correlation_id,
                                                                                              interaction_details,
                                                                                              body,
                                                                                              async_workflow)
-        await self.return_sync_async_response(status, response, wdo)
+        await self.write_response_with_store_updates(status, response, wdo,
+                                                     wd.MessageStatus.OUTBOUND_SYNC_ASYNC_MESSAGE_SUCCESSFULLY_RESPONDED,
+                                                     wd.MessageStatus.OUTBOUND_SYNC_ASYNC_MESSAGE_FAILED_TO_RESPOND)
 
-    async def return_sync_async_response(self, status: int, response: str, wdo: wd.WorkDescription):
+    async def write_response_with_store_updates(self, status: int, response: str, wdo: wd.WorkDescription,
+                                                success, failure):
         try:
             self._write_response(status, response)
-            await wdo.set_outbound_status(wd.MessageStatus.OUTBOUND_SYNC_ASYNC_MESSAGE_SUCCESSFULLY_RESPONDED)
+            if wdo:
+                await wdo.set_outbound_status(success)
         except Exception as e:
             logger.error('0015', 'Failed to respond to supplier system {exception}', {'exception': e})
-            await wdo.set_outbound_status(wd.MessageStatus.OUTBOUND_SYNC_ASYNC_MESSAGE_FAILED_TO_RESPOND)
-
-    async def invoke_default_workflow(self, message_id, correlation_id, interaction_details, body, workflow):
-        status, response = await workflow.handle_outbound_message(message_id,
-                                                                  correlation_id,
-                                                                  interaction_details,
-                                                                  body,
-                                                                  None)
-        self._write_response(status, response)
+            if wdo:
+                await wdo.set_outbound_status(failure)
+
+    async def invoke_default_workflow(self, from_asid, message_id, correlation_id, interaction_details, body, wf):
+        if isinstance(wf, workflow.SynchronousWorkflow):
```

Comment:
I don't like that this is doing different things for the sync workflow and the others. At the very least, you could just wrap the `write_response_with_store_updates` in the check for a sync workflow, and keep the call to `handle_outbound_message` the same for all workflows.

Ideally, we'd push the logic on whether/which states to write once a response is successful into the workflows. Perhaps define an abstract method on the base class that each subclass can provide an implementation of?